### PR TITLE
hal/armv7a/imx6ull: enable MMU for linear addresses translations

### PR DESCRIPTION
**Pull request by HBuczynski**
## Description

- define cached an uncached memories in the linker script
- add MMU controller for `armv7a`
- MMU is enabled for `imx6ull`
- MMU is disabled and cache is flushed and disabled before jump to kernel
- cache management module is added
- as an example, USB buffers are placed in `uncached_ddr`

## Motivation and Context

MMU is enabled to allow allocation of uncached memory.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: ld/armv7a7-imx6ull.ldt [hunk 1]
```diff
@@ -49,9 +59,34 @@ REGION_ALIAS("HEAP", m_ocram_low);
...
+
+/* Additional sections */
+SECTIONS
+{
+	.ddr (NOLOAD) :
+	{
+		. = ALIGN(4);
+		__ddr_start__ = .;
+		*(.ddr*)
+    	__ddr_end__ = .;
```
Comment (gerard5) [inline ld/armv7a7-imx6ull.ldt:77]: Use tabs.
```suggestion
		__ddr_end__ = .;
```

File: ld/armv7a7-imx6ull.ldt [hunk 2]
```diff
@@ -49,9 +59,34 @@ REGION_ALIAS("HEAP", m_ocram_low);
...
+	} > CACHED_DDR
+
+	.uncached_ddr (NOLOAD) :
+	{
+		. = ALIGN(4);
+		__uncached_ddr_start__ = .;
+		*(.uncached_ddr*)
+    	__uncached_ddr_end__ = .;
```
Comment (gerard5) [inline ld/armv7a7-imx6ull.ldt:85]: Use tabs.
```suggestion
		__uncached_ddr_end__ = .;
```

**Review by agkaminski — COMMENTED** [review]
LGTM apart from one thing - IMHO we should disable MMU before entering kernel. There is assumption in kernel that we start with MMU disabled, so some things might be performed wrong or unsafely when we start with MMU enabled. I assume that kernel works anyway, but this might work by accident.

It should be ok to deinit the MMU before exiting to kernel, that's what boot rom is doing

File: hal/armv7a/mmu.c [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+		dsb; \
+		mov r1, #1; \
+		mcr p15, 0, r1, c2, c0, 2; \
+		mcr p15, 0, %0, c2, c0, 0; \
+		mcr p15, 0, %0, c2, c0, 1; \
+		dsb; \
+		isb"
+		:
+		: "r" (addr)
+		: );
```
Comment (kemonats) [inline hal/armv7a/mmu.c:44]:
```suggestion
		:
		: "r" (addr)
		: "r1");
```
Missing `r1` in clobbers.

File: hal/armv7a/mmu.c [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+void mmu_enable(void)
+{
+	__asm__ volatile (" \
+		dsb; \
+		mrc p15, 0, r1, c1, c0, 0; \
+		orr r1, r1, #1; \
+		mcr p15, 0, r1, c1, c0, 0; \
+		dsb; \
+		isb"
+	);
```
Comment (kemonats) [inline hal/armv7a/mmu.c:70]:
```suggestion
		dsb; \
		isb"
		: : : "r1" );
```
Missing `r1` in clobbers.

File: hal/armv7a/mmu.c (hunk 3 above)
Comment (HBuczynski) [reply hal/armv7a/mmu.c]: Thanks, fixed.

**Comment by HBuczynski** [comment]
@agkaminski, I agree with you that the MMU should be disabled before jump to the kernel. I have introduced changes in a new commit. Moreover, the cache memory is flushed and disabled before jump is preformed.

**Review by agkaminski — APPROVED** [review]
LGTM

File: hal/armv7a/_cache.S [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+
+.globl hal_dcacheInvalAll
+.type hal_dcacheInvalAll, %function
+hal_dcacheInvalAll:
+    dsb
+    isb
+    mrc p15, 1, r0, c0, c0, 0        /* Read CCSIDR (Cache Size Identification Register) */
+	mov r3, #0x1ff
+	and r0, r3, r0, lsr #13          /* r0 = number of sets -                            */
```
Comment (gerard5) [inline hal/armv7a/_cache.S:45]: Tabs vs. spaces.

**Review by gerard5 — APPROVED** [review]
Looks good to me.

**Pull request merged**